File: src/services/quota-service.js
/**
 * 生成配额：代理模式走服务端，本地模式走 localStorage
 */
import { api } from './api-client.js';
import { Storage } from '../core/storage.js';
import { DEFAULT_SETTINGS } from '../core/config.js';
import { logger } from '../core/logger.js';

const LOCAL_KEY = 'cleartalk_daily_generations';

function todayKey() {
  return new Date().toISOString().slice(0, 10);
}

function defaultLocalQuota() {
  const free = 50;
  const raw = Storage.getSetting('localDailyLimit', free);
  const limit = parseInt(raw, 10) || free;
  const data = JSON.parse(localStorage.getItem(LOCAL_KEY) || '{}');
  const used = data[todayKey()] || 0;
  return {
    tier: 'free',
    limit,
    used,
    remaining: Math.max(0, limit - used),
    resetsAt: new Date(new Date().setHours(23, 59, 59, 999)).toISOString(),
    source: 'local'
  };
}

function consumeLocal() {
  const data = JSON.parse(localStorage.getItem(LOCAL_KEY) || '{}');
  const key = todayKey();
  data[key] = (data[key] || 0) + 1;
  localStorage.setItem(LOCAL_KEY, JSON.stringify(data));
}

export const QuotaService = {
  async getStatus() {
    const provider = Storage.getSetting('aiProvider', DEFAULT_SETTINGS.aiProvider);
    if (provider !== 'proxy') {
      return defaultLocalQuota();
    }
    try {
      const status = await api.get('/api/quota');
      return { ...status, source: 'remote' };
    } catch (err) {
      logger.warn('Quota fetch failed:', err.message);
      return defaultLocalQuota();
    }
  },

  async checkBeforeGenerate() {
    const status = await this.getStatus();
    if (status.remaining <= 0) {
      const err = new Error('QUOTA_EXCEEDED');
      err.quota = status;
      throw err;
    }
    return status;
  },

  async recordGenerate() {
    const provider = Storage.getSetting('aiProvider', DEFAULT_SETTINGS.aiProvider);
    if (provider !== 'proxy') {
      consumeLocal();
      return;
    }
    // 服务端在 generate 成功时已扣减；此处仅刷新缓存
  }
};

export default QuotaService;

File: lib/services/quota_service.dart
import 'package:dio/dio.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'proxy_ai_service.dart';

class QuotaStatus {
  final String tier;
  final int limit;
  final int used;
  final int remaining;

  QuotaStatus({
    required this.tier,
    required this.limit,
    required this.used,
    required this.remaining,
  });

  factory QuotaStatus.fromJson(Map<String, dynamic> json) {
    return QuotaStatus(
      tier: json['tier']?.toString() ?? 'free',
      limit: (json['limit'] as num?)?.toInt() ?? 20,
      used: (json['used'] as num?)?.toInt() ?? 0,
      remaining: (json['remaining'] as num?)?.toInt() ?? 0,
    );
  }
}

class QuotaService {
  final Dio _dio;

  QuotaService({Dio? dio}) : _dio = dio ?? Dio();

  Future<QuotaStatus> fetch() async {
    final prefs = await SharedPreferences.getInstance();
    final base = prefs.getString('cleartalk_api_base_url')?.trim() ??
        ProxyAiService.defaultBaseUrl;
    try {
      final res = await _dio.get('$base/api/quota');
      return QuotaStatus.fromJson(Map<String, dynamic>.from(res.data));
    } catch (_) {
      return QuotaStatus(tier: 'free', limit: 50, used: 0, remaining: 50);
    }
  }
}
